File: trivially/src/main/java/trivially/ep/m6/Mult.java
package trivially.ep.m6;

public interface Mult extends Exp, trivially.ep.m5.Mult {

    Exp getLeft();

    Exp getRight();

    default Boolean equals(Exp other) {
        return this.astree().equals(other.astree());
    }

    default Exp simplify() {
        if (this.getLeft().eval().equals(0.0) || this.getRight().eval().equals(0.0)) {
            return new trivially.ep.m6.finalized.Lit(0.0);
        } else if (this.getLeft().eval().equals(1.0)) {
            return this.getRight().simplify();
        } else if (this.getRight().eval().equals(1.0)) {
            return this.getLeft().simplify();
        } else {
            return new trivially.ep.m6.finalized.Mult(this.getLeft().simplify(), this.getRight().simplify());
        }
    }

    default Boolean isMult(Exp left, Exp right) {
        return left.eql(getLeft()) && right.eql(getRight());
    }

    default Boolean eql(Exp that) { return that.isMult(getLeft(), getRight()); }

    default void truncate (int level) {
        if (level > 1) {
            getLeft().truncate(level-1);
            getRight().truncate(level-1);
        } else {
            setLeft(new trivially.ep.m5.finalized.Lit(getLeft().eval()));
            setRight(new trivially.ep.m5.finalized.Lit(getRight().eval()));
        }
    }
}

File: coco_csharp/CoCo/EP/M6/Mult.cs
namespace EP.M6 {
    public interface Mult<FT> : Exp<FT>, EP.M5.Mult<FT> {
        bool Exp<FT>.equals(EP.Exp<FT> other) {
            return convert(this).astree().Equals(convert(other).astree());
        }

        bool Exp<FT>.isMult(EP.Exp<FT> left, EP.Exp<FT> right) {
            return convert(left).eql(getLeft()) && convert(right).eql(getRight());
        }

        bool Exp<FT>.eql(EP.Exp<FT> that) {
            return convert(that).isMult(convert(getLeft()), convert(getRight()));
        }
    }
}